### PR TITLE
Refactor value source

### PR DESCRIPTION
**Pull request by kaapstorm**
##### SUMMARY
Value sources are used in MOTECH for mapping CommCare values to remote API fields. Value sources define where MOTECH gets values from, like form questions, or case properties,or  location metadata.

The primary objective of this PR is to make defining value sources simpler for third-party administrators.

The secondary objective is to simplify the code around value sources.

Currently, the `ValueSource` class ...
* subclasses `DocumentSchema`.
* has many subclasses.
* uses the value of its "doc_type" attribute to determine which subclass the user needs.

With this refactor, `ValueSource` ...
* is an `attrs` class.
* has few subclasses.
* determines which subclass the user needs by checking which properties the user wants to use.
* works with all current configurations, so no migration is required before this PR can be deployed.

Review :fish: by :tropical_fish: commit :blowfish: 

Updates to documentation will be covered in a follow-up PR.

##### FEATURE FLAG
"OpenMRS integration" or "DHIS2 integration"

##### PRODUCT DESCRIPTION
Simplifies how integrations are configured -- specifically how CommCare values are mapped to external values for DHIS2 and OpenMRS.

### CODE REVIEW

**Comment by orangejenny** [comment]
Killed tests so I can deploy something else, will restart them once that's done.

File: corehq/motech/value_source.py [hunk 1]
```diff
@@ -253,160 +246,186 @@ class ConstantValue(ConstantString):
...
-class JsonPathCasePropertyMap(CasePropertyMap, JsonPathMixin):
-    pass
+def as_value_source(data: dict) -> ValueSource:
+    for subclass in recurse_subclasses(ValueSource):
+        try:
+            args, kwargs = subclass.get_schema_params()
+            # .. WARNING::
+            #    ``validate()`` can modify ``data``. This can break
+            #    validation for other subclasses! If it is ever
+            #    necessary to modify ``data``, it should be deep-copied
```
Comment (millerdev) [inline corehq/motech/value_source.py]: I'm not sure how to reconcile "_``validate()`` can modify ``data``_" with "_If it is ever necessary to modify ``data``, ..._".

Is the phrase "_``validate()`` can modify ``data``_" stating that it is a possibility given some theoretical implementation of `validate()` that does not currently exist at the time of writing this comment? Or does it mean there are known implementations that do modify `data`? If the latter then this should be doing a deep copy now, right?

**Review by millerdev — APPROVED** [review]
I like the change to make `ValueSource` use `attr.s`. But I feel like I've reviewed a fair amount of this a few times before, and my eyes my be glazing over a bit. I hope I caught everything.

File: corehq/motech/value_source.py (hunk 1 above)
Comment (kaapstorm) [reply corehq/motech/value_source.py]: > does it mean there are known implementations that do modify data?

That is what I thought. But I thought wrong! I reread the code, and `validate()` returns a new dictionary.

> If the latter then this should be doing a deep copy now, right?

I don't love deep copy. My reason for that (incorrect) comment was to tell a future developer to add `deepcopy()` only when they wrote a schema that modified values. I agree that adding it pre-emptively would future-proof the code. And considering all ValueSource classes are shallow and small, `deepcopy()` would be fast, so speed and resources wouldn't be an issue.

Luckily `validate()` is sensible, and doesn't modify its parameter by reference. So I've added a test to verify that, and I'm dropping the comment.

Comment (millerdev) [reply corehq/motech/value_source.py]: > I don't love deep copy.

I agree. It should be avoided. I'm glad you were able to invalidate and remove that comment.

**Comment by millerdev** [comment]
🎉

**Pull request merged**